lib/matplotlib/__init__.py

```diff
@@ -1515,6 +1515,43 @@ def test(verbosity=1):
 
 test.__test__ = False  # nose: this function is not a test
 
+
+def _replacer(data, key):
+    # if key isn't a string don't bother
+    if not isinstance(key, six.string_types):
+        return key
+    # try to use __getitem__
+    try:
+        return data[key]
+    # key does not exist, silently fall back to key
+    except KeyError:
+        return key
+
+
+def unpack_labeled_data(func):
+    """
+    A decorator to add a 'data' kwarg to any a function.  The signature
+    of the input function must be ::
+
+       def foo(ax, *args, **kwargs)
+
+    so this is suitable for use with Axes methods.
+    """
+    @functools.wraps(func)
+    def inner(ax, *args, **kwargs):
+        data = kwargs.pop('data', None)
+        if data is not None:
+            if rcParams['unpack_labeled']:
+                args = tuple(_replacer(data, a) for a in args)
+                kwargs = dict((k, _replacer(data, v))
+                              for k, v in six.iteritems(kwargs))
+            else:
+                raise ValueError("Trying to unpack labeled data, but "
+                                 "rcParams['unpack_labeled'] is False")
+
+        return func(ax, *args, **kwargs)
+    return inner
```

Comment:
In the long run, is this going to need to add something to the docstring as well, so that this unpacking is part of the API as described in the docstring?  Somehow, very concisely described?  Maybe via a reference?